wgpu-hal/src/dx12/device.rs

```diff
@@ -221,62 +222,148 @@ impl super::Device {
             .map(|name| ffi::CString::new(name.as_str()).unwrap())
             .map_err(|e| crate::PipelineError::Linkage(stage_bit, format!("{}", e)))?;
 
-        let mut shader_data = native::Blob::null();
-        let mut error = native::Blob::null();
-        let mut compile_flags = d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS;
+        let source_name = stage
+            .module
+            .raw_name
+            .as_ref()
+            .and_then(|cstr| cstr.to_str().ok())
+            .unwrap_or_default();
+
+        let mut compile_flags = vec!["-Ges"]/* d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS */;
         if self
             .private_caps
             .instance_flags
             .contains(crate::InstanceFlags::DEBUG)
         {
-            compile_flags |=
-                d3dcompiler::D3DCOMPILE_DEBUG | d3dcompiler::D3DCOMPILE_SKIP_OPTIMIZATION;
+            compile_flags.push("-Zi"); /* d3dcompiler::D3DCOMPILE_SKIP_OPTIMIZATION */
+            compile_flags.push("-Od"); /* d3dcompiler::D3DCOMPILE_DEBUG */
         }
 
-        let source_name = match stage.module.raw_name {
-            Some(ref cstr) => cstr.as_c_str().as_ptr(),
-            None => ptr::null(),
+        let load_shader_fxc = || {
+            let mut shader_data = native::Blob::null();
+            let mut compile_flags = d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS;
+
+            if self
+                .private_caps
+                .instance_flags
+                .contains(crate::InstanceFlags::DEBUG)
+            {
+                compile_flags |=
+                    d3dcompiler::D3DCOMPILE_DEBUG | d3dcompiler::D3DCOMPILE_SKIP_OPTIMIZATION;
+            }
+
+            let full_stage = format!(
+                "{}_{}\0",
+                naga_stage.to_hlsl_str(),
+                hlsl::ShaderModel::V5_1.to_str()
+            );
+            let mut error = native::Blob::null();
+
+            let hr = unsafe {
+                profiling::scope!("d3dcompiler::D3DCompile");
+                d3dcompiler::D3DCompile(
+                    source.as_ptr() as *const _,
```

Comment:
Preferable to use .cast() here instead of as casts for all of these pointer casts.